test/release/examples/patterns/readcsv.chpl

```diff
@@ -0,0 +1,286 @@
+
+  var reader = f.reader();
+
+  // Read the first line to get the column names.
+  // Assuming the following format, where the number of columns (ncol) is not 
+  // known ahead of time:
+  //    col_name_1,col_name_2,...,col_name_ncol\n
+  var line : string;
+
+  if (!reader.readline(line)) then
+    writeln("ERROR: ", inFileName, " appears to be empty");
+
+  var colNames : list(string);
+  colNames = createListOfColNames(line);
+
+  if debug {
+    writeln("\ncolNames: ", colNames);
+  }
+
+  // Create a list of maps with one map per row of data.
+  var dataRows : list(map(string,string));
+  // Read individual lines from the file until the end of the file.
+  while (reader.readline(line)) {
+    // Create a new map per line.
+    var aRowMap = new map(string, string);
+    var start = 0;      // index into the line string
+    var colIdx = 0;     // the first column is indexed as zero
+    // Process each line by finding the next field value from the start index to the next comma.
+    while (start<line.size && colIdx<colNames.size) {
+      var (nextVal,nextCommaIdx) = nextField(line,start);
+      aRowMap.add(colNames[colIdx],nextVal);
+      start = nextCommaIdx + 1;         // The next column of data will start after the comma.
+      colIdx = colIdx + 1;              // Increment the column index.
+    }
+    dataRows.append(aRowMap);
+  }
+  reader.close();
+
+  // Print out the last data row to illustrate what one of the row maps contains.
+  if debug then writeln("\ndataRows[dataRows.size-1] = ", dataRows[dataRows.size-1]);
+ 
+  //-----------------------------------------------------------------------
+  // To see how the data is accessed and for testing, let's generate a 
+  // temporary csv file from the stored data.  It should be the same as the 
+  // input csv file.
+  if debug {
+    var outfile = open(approach1FileName, iomode.cw);
+    var writer = outfile.writer();
+
+    // First write to the output file the column names separated by commas.
+    for colIdx in 0..colNames.size-2 {
+      writer.write(colNames[colIdx],",");
+    }
+    // The last column name should not be followed by comma.
+    writer.writeln(colNames[colNames.size-1]);
+
+    // Then write to the file each row of data.
+    for row in dataRows {
+      for colIdx in 0..colNames.size-2 {
+        writer.write(row[colNames[colIdx]],",");
+      }
+      // The last column value should not be followed by comma.
+      writer.writeln(row[colNames[colNames.size-1]]);
+    }
+    writer.close();
+
+    // Compare the file generated from the list of maps with the original input csv file.
+    checkThatFilesMatch(inFileName,approach1FileName);
+  }
+}
+
+// Approach 2: chapeltastic -- create an associative array with the key being the 
+//             column name and the value being an array of strings
+{
+  writeln();
+  writeln();
+  writeln("Approach 2: associative array keyed of column name with array of vals per column"); 
```

Comment:
I find the description a bit hard to understand. Maybe something like "associative array where keys are column names, and values are arrays of row values"

Reply:
Good idea. Rewrote as you suggested.